src/psyclone/psyir/nodes/routine.py

```diff
@@ -0,0 +1,211 @@
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author: A. R. Porter, STFC Daresbury Lab
+# -----------------------------------------------------------------------------
+
+''' This module contains the Routine node implementation.'''
+
+from psyclone.psyir.nodes.schedule import Schedule
+from psyclone.psyir.symbols import DataType
+from psyclone.psyir.nodes.node import Node
+from psyclone.psyir.symbols.symboltable import SymbolTable
+from psyclone.errors import GenerationError
+
+
+class Routine(Schedule):
+    '''
+    A sub-class of a Schedule that represents a subroutine, function or
+    program unit.
+
+    :param str name: the name of this routine.
+    :param bool is_program: whether this Routine represents the entry point \
+                            into a program (i.e. Fortran Program or C main()).
+    :param return_type: the return-type of this routine.
+    :type return_type: :py:class:`psyclone.psyir.symbols.DataType` or NoneType
+    :param children: the PSyIR nodes that are children of this Routine.
+    :type children: list of :py:class:`psyclone.psyir.nodes.Node`
+    :param parent: the parent node of this Schedule in the PSyIR.
+    :type parent: :py:class:`psyclone.psyir.nodes.Node` or NoneType
+    :param symbol_table: initialise the Schedule with a given symbol table.
+    :type symbol_table: :py:class:`psyclone.psyir.symbols.SymbolTable` or \
+            NoneType
+
+    :raises TypeError: if the supplied routine name is not a str.
+    :raises TypeError: if the supplied is_program is not a bool.
+    :raises TypeError: if the supplied return_type is not a DataType.
+
+    '''
+    # Textual description of the node.
+    _children_valid_format = "[Statement]*"
+    _text_name = "Routine"
+    _colour_key = "Schedule"
+
+    def __init__(self, name, is_program=False, return_type=None,
+                 children=None, parent=None, symbol_table=None):
+        super(Routine, self).__init__(children=children, parent=parent,
+                                      symbol_table=symbol_table)
+        self.name = name
+
+        if not isinstance(is_program, bool):
+            raise TypeError("Routine 'is_program' must be a bool but got "
+                            "'{0}'".format(type(is_program).__name__))
+        self._is_program = is_program
+
+        if return_type and not isinstance(return_type, DataType):
+            raise TypeError("Routine 'return_type' must be of type DataType "
+                            "but got '{0}'".format(type(return_type).__name__))
+        self._return_type = return_type
+
+    @classmethod
```

Comment:
Perhaps add a comment explaining why classmethod is used i.e. to allow this method to be called from a classmethod in a subclass and work as expected.

Reply:
Done.